waspc/cli/Command/Watch.hs

```diff
@@ -35,19 +44,45 @@ watch waspProjectDir outDir = FSN.withManager $ \mgr -> do
   chan <- newChan
   _ <- FSN.watchDirChan mgr (SP.toFilePath waspProjectDir) eventFilter chan
   _ <- FSN.watchTreeChan mgr (SP.toFilePath $ waspProjectDir </> Common.extCodeDirInWaspProjectDir) eventFilter chan
-  listenForEvents chan currentTime
+  timeOfLastEvent <- MVar.newMVar currentTime
+  listenForEvents chan currentTime timeOfLastEvent
   where
-    listenForEvents :: Chan FSN.Event -> UTCTime -> IO ()
-    listenForEvents chan lastCompileTime = do
+    oneSecond :: Int
+    oneSecond = 1000000
+
+    oneSecondDelay :: MVar.MVar UTCTime -> IO ()
+    oneSecondDelay timeOfLastEventMVar = do
+      threadDelay oneSecond
+      currentDelayTime <- MVar.readMVar timeOfLastEventMVar
+      currentTime <- getCurrentTime
+      let timeDiff = nominalDiffTimeToSeconds $ diffUTCTime currentTime currentDelayTime
+      when (timeDiff < fromIntegral oneSecond) $ do
+        threadDelay (oneSecond - (floor . (* 1e9) $ timeDiff))
+        oneSecondDelay timeOfLastEventMVar
+
+    recurEventCheck :: Chan FSN.Event -> UTCTime -> MVar.MVar UTCTime -> IO ()
+    recurEventCheck chan currentTime timeOfLastEventMVar = do
+      event <- readChan chan
+      let eventTime = FSN.eventTime event
+      unless (eventTime < currentTime) $ do
+        MVar.putMVar timeOfLastEventMVar eventTime
+
+    listenForEvents :: Chan FSN.Event -> UTCTime -> MVar.MVar UTCTime -> IO ()
+    listenForEvents chan lastCompileTime timeOfLastEventMVar = do
```

Comment:
I don't think you need to have `timeOfLastEventMVar` here -> listenForEvents function does not need to take it from outside. timeOfLastEvent is important only while waiting for that one second of time between the events, which happens inside the `listenForEvents`, opaque to the external caller.

So, you should create this MVar in the `listenForEvents` just before you start the `race` and pass it down to recurEventCheck and oneSecondDelay.